src/mobu/services/business/notebookrunner.py

```diff
@@ -295,26 +315,88 @@ async def execute_code(self, session: JupyterLabSession) -> None:
             msg = f"Notebook {self._notebook.name} iteration {iteration}"
             self.logger.info(msg)
 
-            for cell in self.read_notebook(self._notebook):
-                code = "".join(cell["source"])
-                cell_id = cell.get("id") or cell["_index"]
-                ctx = CodeContext(
-                    notebook=self._notebook.name,
-                    path=str(self._notebook),
-                    cell=cell_id,
-                    cell_number=f"#{cell['_index']}",
-                    cell_source=code,
-                )
-                await self.execute_cell(session, code, cell_id, ctx)
-                if not await self.execution_idle():
-                    break
+            with self.timings.start(
+                "execute_notebook", self.annotations(self._notebook.name)
+            ) as sw:
+                try:
+                    for cell in self.read_notebook(self._notebook):
+                        code = "".join(cell["source"])
+                        cell_id = cell.get("id") or cell["_index"]
+                        ctx = CodeContext(
+                            notebook=self._notebook.name,
+                            path=str(self._notebook),
+                            cell=cell_id,
+                            cell_number=f"#{cell['_index']}",
+                            cell_source=code,
+                        )
+                        await self.execute_cell(session, code, cell_id, ctx)
+                        if not await self.execution_idle():
+                            break
+                except:
+                    await self._publish_notebook_failure()
+                    raise
 
             self.logger.info(f"Success running notebook {self._notebook.name}")
+            await self._publish_notebook_success(sw.elapsed)
             if not self._notebook_paths:
                 self.logger.info("Done with this cycle of notebooks")
             if self.stopping:
                 break
 
+    async def _publish_notebook_success(self, duration: timedelta) -> None:
+        await ed.events.notebook_execution.publish(
+            NotebookExecution(
+                **self.common_notebook_event_attrs(),
+                duration=duration,
+                success=True,
+            )
+        )
+
+    async def _publish_notebook_failure(self) -> None:
+        await ed.events.notebook_execution.publish(
+            NotebookExecution(
+                **self.common_notebook_event_attrs(),
+                duration=None,
```

Comment:
We definitely want to record duration for failure as well. That's been the only way that I've been able to figure out that the failure was due to a timeout.